c_src/vips_image.c

```diff
@@ -220,6 +220,40 @@ ERL_NIF_TERM nif_image_write_to_buffer(ErlNifEnv *env, int argc,
   return ret;
 }
 
+ERL_NIF_TERM nif_image_write_to_array(ErlNifEnv *env, int argc,
+                                       const ERL_NIF_TERM argv[]) {
+  ASSERT_ARGC(argc, 1);
+
+  VipsImage *image;
+  ErlNifTime start;
+  ERL_NIF_TERM ret;
+  void *array;
+  size_t size;
+
+  start = enif_monotonic_time(ERL_NIF_USEC);
+
+  if (!erl_term_to_g_object(env, argv[0], (GObject **)&image)) {
+    ret = make_error(env, "Failed to get VipsImage");
+    goto exit;
+  }
+
+  array = vips_image_write_to_memory(image, &size);
+
+  if (!array) {
+    error("Failed to write VipsImage to array. error: %s",
+          vips_error_buffer());
+    vips_error_clear();
+    ret = make_error(env, "Failed to write VipsImage to array");
+    goto exit;
+  }
+
+  ret = make_ok(env, g_object_to_erl_term(env, (GObject *)array));
```

Comment:
`array` in this case it is a pointer to a raw memory block, so we can't treat it like [g_object](https://docs.gtk.org/gobject/concepts.html#the-gobject-base-class). 

There are a couple of options here, we can convert it to binary and transfer the ownership, or we can use `GBoxed` (or libvips [abstraction](https://www.libvips.org/API/current/libvips-type.html#VipsArea) over it), or build our own solution. We also need to consider how are we going to support concurrency with multiple schedulers accessing the same memory.